File: frontend/src/lib/queries/weight_performance.ts
// SP04 Stage 4c — server-only queries against the live-perf + revert
// audit tables. Used by /admin/weight-performance and by the revert
// banner on /admin/composite-proposals.
import 'server-only'
import sql from '@/lib/db'

export type LivePerfRow = {
  weight_set_version: string
  as_of_date: Date
  realized_ic: string
  ic_ratio: string | null
  n_observations: number
}

export type ActiveSetSummary = {
  tier: string
  regime: string
  version: string
  predicted_ic: string | null
  trail: LivePerfRow[]
  days_below_threshold: number
  n_trail_rows: number
  in_revert_territory: boolean
}

export type RevertLogRow = {
  id: string
  tier: string
  regime: string
  reverted_from_version: string
  restored_to_version: string | null
  days_below_threshold: number
  realized_ic_avg: string | null
  predicted_holdout_ic: string | null
  triggered_by: string
  notes: string | null
  applied_at: Date
}

export type HitRateRow = {
  instrument_id: string
  date: Date
  lookback_window: number
  n_high_conviction_days: number
  n_positive_outcomes: number
  hit_rate: string | null
  tier_at_as_of: string | null
}

export async function getActiveWeightSetsWithTrail(): Promise<ActiveSetSummary[]> {
  // Pull active sets + their trails in one round trip via two queries.
  type ActiveRow = {
    tier: string
    regime: string
    version: string
    predicted_ic: string | null
  }
  const actives = await sql<ActiveRow[]>`
    SELECT tier, regime,
           tier || '@' || MAX(approved_at)::text AS version,
           MAX(holdout_ic)::text                  AS predicted_ic
    FROM atlas.atlas_signal_weights
    WHERE effective_to IS NULL
    GROUP BY tier, regime
    ORDER BY tier
  `
  const out: ActiveSetSummary[] = []
  for (const a of actives) {
    const trail = await sql<LivePerfRow[]>`
      SELECT
        weight_set_version,
        as_of_date,
        realized_ic::text AS realized_ic,
        ic_ratio::text    AS ic_ratio,
        n_observations
      FROM atlas.atlas_signal_weights_live_perf
      WHERE weight_set_version = ${a.version}
        AND as_of_date >= CURRENT_DATE - INTERVAL '30 days'
      ORDER BY as_of_date ASC
    `
    const daysBelow = trail.filter(
      (t) => t.ic_ratio !== null && parseFloat(t.ic_ratio) < 0.5,
    ).length
    out.push({
      tier: a.tier,
      regime: a.regime,
      version: a.version,
      predicted_ic: a.predicted_ic,
      trail,
      days_below_threshold: daysBelow,
      n_trail_rows: trail.length,
      in_revert_territory: trail.length >= 60 && daysBelow === trail.length,
    })
  }
  return out
}

export async function getRecentReverts(): Promise<RevertLogRow[]> {
  return await sql<RevertLogRow[]>`
    SELECT
      id::text                  AS id,
      tier,
      regime,
      reverted_from_version,
      restored_to_version,
      days_below_threshold,
      realized_ic_avg::text     AS realized_ic_avg,
      predicted_holdout_ic::text AS predicted_holdout_ic,
      triggered_by,
      notes,
      applied_at
    FROM atlas.atlas_weight_revert_log
    WHERE applied_at >= NOW() - INTERVAL '30 days'
    ORDER BY applied_at DESC
  `
}

export async function getHitRateForStock(
  instrumentId: string,
  lookback: number = 20,
): Promise<HitRateRow | null> {
  const rows = await sql<HitRateRow[]>`
    SELECT
      instrument_id::text       AS instrument_id,
      date,
      lookback_window,
      n_high_conviction_days,
      n_positive_outcomes,
      hit_rate::text            AS hit_rate,
      tier_at_as_of
    FROM atlas.atlas_stock_hit_rate_daily
    WHERE instrument_id = ${instrumentId}::uuid
      AND lookback_window = ${lookback}
    ORDER BY date DESC
    LIMIT 1
  `
  return rows[0] ?? null
}

File: atlas/api/admin/weight_performance.py
"""Stage 4c — admin endpoint: live performance + recent reverts.

GET /api/admin/weight-performance returns, for each currently-active
weight set, its predicted IC and a 30-day rolling realized-IC trail,
plus the count of recent days below the 0.5 ratio threshold. Also lists
the most recent revert events (last 30 days).

Auth: same dual-path (_require_admin) as the proposals route — JWT
role=admin or ATLAS_INTERNAL_SECRET bearer.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

import structlog
from fastapi import APIRouter, Request
from sqlalchemy import text

from atlas.api.admin.proposals import _require_admin
from atlas.db import get_engine

log = structlog.get_logger()

router = APIRouter(prefix="/api/admin/weight-performance", tags=["admin"])


_ACTIVE_VERSIONS_SQL = text("""
    SELECT tier, regime,
           tier || '@' || MAX(approved_at)::text AS version,
           MAX(holdout_ic) AS predicted_ic
    FROM atlas.atlas_signal_weights
    WHERE effective_to IS NULL
    GROUP BY tier, regime
    ORDER BY tier
""")

_RECENT_PERF_SQL = text("""
    SELECT weight_set_version, as_of_date,
           realized_ic, ic_ratio, n_observations
    FROM atlas.atlas_signal_weights_live_perf
    WHERE weight_set_version = :version
      AND as_of_date >= CURRENT_DATE - INTERVAL '30 days'
    ORDER BY as_of_date ASC
""")

_RECENT_REVERTS_SQL = text("""
    SELECT id::text, tier, regime,
           reverted_from_version, restored_to_version,
           days_below_threshold, realized_ic_avg, predicted_holdout_ic,
           triggered_by, notes, applied_at
    FROM atlas.atlas_weight_revert_log
    WHERE applied_at >= NOW() - INTERVAL '30 days'
    ORDER BY applied_at DESC
""")


@router.get("")
async def list_weight_performance(request: Request) -> dict[str, Any]:
    _require_admin(request)
    engine = get_engine()

    with engine.connect() as conn:
        actives = conn.execute(_ACTIVE_VERSIONS_SQL).fetchall()

        active_sets: list[dict[str, Any]] = []
        for row in actives:
            tier, regime, version, predicted = row[0], row[1], row[2], row[3]
            perf_rows = conn.execute(_RECENT_PERF_SQL, {"version": version}).fetchall()
            trail: list[dict[str, Any]] = [
                {
                    "date": str(p[1]),
                    "realized_ic": float(p[2]) if p[2] is not None else None,
                    "ratio": float(p[3]) if p[3] is not None else None,
                    "n_observations": int(p[4]),
                }
                for p in perf_rows
            ]
            days_below = 0
            for t in trail:
                r = t.get("ratio")
                if isinstance(r, (int, float)) and r < 0.5:  # noqa: UP038
                    days_below += 1
            active_sets.append(
                {
                    "tier": tier,
                    "regime": regime,
                    "version": version,
                    "predicted_ic": float(predicted) if predicted is not None else None,
                    "trail": trail,
                    "days_below_threshold": days_below,
                    "n_trail_rows": len(trail),
                    "in_revert_territory": days_below == len(trail) and len(trail) >= 60,
                }
            )

        revert_rows = conn.execute(_RECENT_REVERTS_SQL).fetchall()

    reverts = [
        {
            "id": r[0],
            "tier": r[1],
            "regime": r[2],
            "reverted_from": r[3],
            "restored_to": r[4],
            "days_below_threshold": int(r[5]),
            "realized_ic_avg": float(r[6]) if r[6] is not None else None,
            "predicted_holdout_ic": float(r[7]) if r[7] is not None else None,
            "triggered_by": r[8],
            "notes": r[9],
            "applied_at": r[10].isoformat() if isinstance(r[10], datetime) else str(r[10]),
        }
        for r in revert_rows
    ]

    return {"active_sets": active_sets, "recent_reverts": reverts}
